lib/compat.php

```diff
@@ -24,18 +24,90 @@
 	 * @return WP_Block_Type|false The registered block type on success, or false on failure.
 	 */
 	function register_block_type_from_metadata( $path, $args = array() ) {
-		$file = trailingslashit( $path ) . 'block.json';
-		if ( ! file_exists( $file ) ) {
+		$metadata_file = trailingslashit( $path ) . 'block.json';
+		if ( ! file_exists( $metadata_file ) ) {
 			return false;
 		}
 
-		$metadata = json_decode( file_get_contents( $file ), true );
-		if ( ! is_array( $metadata ) ) {
+		$metadata = json_decode( file_get_contents( $metadata_file ), true );
+		if ( ! is_array( $metadata ) || ! $metadata['name'] ) {
 			return false;
 		}
 
+		$block_dir           = dirname( $metadata_file );
+		$block_name          = $metadata['name'];
+		$asset_handle_prefix = str_replace( '/', '-', $block_name );
+
+		if ( isset( $metadata['editorScript'] ) ) {
+			$editor_script            = $metadata['editorScript'];
+			$editor_script_handle     = "$asset_handle_prefix-editor-script";
+			$editor_script_asset_path = realpath(
+				$block_dir . '/' . substr_replace( $editor_script, '.asset.php', -3 )
+			);
+			if ( ! file_exists( $editor_script_asset_path ) ) {
+				throw new Error(
+					"The asset file for the \"editorScript\" defined in \"$block_name\" block definition is missing."
+				);
+			}
+			$editor_script_asset  = require( $editor_script_asset_path );
+			wp_register_script(
+				$editor_script_handle,
+				plugins_url( $editor_script, $metadata_file ),
+				$editor_script_asset['dependencies'],
+				$editor_script_asset['version']
+			);
+			unset( $metadata['editorScript'] );
+			$metadata['editor_script'] = $editor_script_handle;
+		}
+
+		if ( isset( $metadata['script'] ) ) {
+			$script            = $metadata['script'];
+			$script_handle     = "$asset_handle_prefix-script";
+			$script_asset_path = realpath(
+				$block_dir . '/' . substr_replace( $script, '.asset.php', -3 )
+			);
+			if ( ! file_exists( $script_asset_path ) ) {
+				throw new Error(
+					"The asset file for the \"script\" defined in \"$block_name\" block definition is missing."
+				);
+			}
+			$script_asset  = require( $script_asset_path );
+			wp_register_script(
+				$script_handle,
+				plugins_url( $script, $metadata_file ),
+				$script_asset['dependencies'],
+				$script_asset['version']
+			);
+			$metadata['script'] = $script_handle;
+		}
+
+		if ( isset( $metadata['editorStyle'] ) ) {
+			$editor_style        = $metadata['editorStyle'];
+			$editor_style_handle = "$asset_handle_prefix-editor-style";
```

Comment:
In RFC we mentioned usage of asset files for CSS, do we really need it? It's definitely something that could be conisdered as a follow-up once CSS imports for build tools are included in `@wordpress/scripts` as proposed in https://github.com/WordPress/gutenberg/pull/21730.

Reply:
> In RFC we mentioned usage of asset files for CSS, do we really need it?

I'm not sure how "appetizing" using `filemtime` as the version of the stylesheet will be as a formal solution to style registration.

Reply:
I copied it over from the template that Create Block uses that was probably done based on this tutorial 😅:
https://developer.wordpress.org/block-editor/tutorials/block-tutorial/applying-styles-with-stylesheets/

In https://developer.wordpress.org/reference/functions/wp_register_style/#parameters it describes `$ver` as:

> (`string`|`bool`|`null`) (Optional) String specifying stylesheet version number, if it has one, which is added to the URL as a query string for cache busting purposes. If version is set to `false`, a version number is automatically added equal to current installed WordPress version. If set to `null`, no version is added.
Default value: `false`

I don't think `false` is optimal in this case because plugins/blocks can change without version update for WordPress core. Should we make it possible to use plugin's `version` in `block.json` that would be the primary way to maintan browser's cache?

Reply:
> I don't think `false` is optimal in this case because plugins/blocks can change without version update for WordPress core. Should we make it possible to use plugin's `version` in `block.json` that would be the primary way to maintan browser's cache?

I'd agree that it wouldn't be great to not be passing any version here at all. The plugin version _could_ work, though it's a bit more magical than is common. For example, a plugin implementer still needs to manually specify a version for any scripts they register, even if technically core could infer it from the plugin manifest.

I guess I was mostly thinking in terms of the question posed, since the problem is solved for scripts by using the generated hash from the asset file. If we had similarly for CSS here, it should address the issue.

In the meantime, `filemtime` might be a fine enough solution. At worst, it invalidates the cache more frequently than is necessary.

Reply:
It makes me wonder whether we could bring handling of styles closer to scripts by generating the asset file as well. The challenge here is that it’s less likely that CSS is managed with webpack and it would be much harder to come up with a strategy that works for the majority of use cases.

Reply:
I listed it in the description as a follow-up task.